### PR TITLE
feat: teach `kubit local apply` to use containers

### PR DESCRIPTION
**Pull request by jdockerty**
By moving to containers for this output, we do not rely on the user running the `kubit local apply` to have the latest (or specific) version of our tools that we rely on, e.g. this decouples the need for `kubecfg` to be at a specific version to ensure that reliant functionality works. We can simply pin the relevant versions of tools alongside `kubit` itself.

### CODE REVIEW

File: src/apply.rs
Comment (jdockerty) [inline src/apply.rs:7]: Introducing this `is_local` in a bunch of places feels a bit heavy-handed, but my intention is to get it something working first and then refactor from a known good point.

File: src/apply.rs
Comment (jdockerty) [inline src/apply.rs:35]: I'm not a huge fan of this `if-else` branching here, but I think the "use containers" approach is a little more complicated than I first thought.

If we don't branch between local and running controller, we are then using "docker-in-docker". 

Perhaps we instead mount the Docker socket though and we're okay with that? Meaning we have sibling containers, as opposed to child containers (acts like d-in-d, but isn't d-in-d).

This is a part where I might've over-complicated it, I think if this is simplified then the overall approach may also be easier to implement too.

**Review by jdockerty — COMMENTED** [review]
Checkpoint self-review

Ignore the failing CI, I know that `clippy` is complaining about stuff but I'm using this to get feedback on the approach rather than digging myself into a huge hole :smile:

File: src/apply.rs
Comment (mkmik) [reply src/apply.rs:35]: most flags are shared. You can use a mutable vector, populate it with the right command depending on the is_local, and then adding the flags that are common between the two kinds of invocations

`let mut cli = Vec::new()`

and then `cli.extend(...)`

Comment (jdockerty) [reply src/apply.rs:35]: I'll have to take another look at this, I might've messed it up initially.

This was my first intention, but using `.extend` was similar to an appending of arguments, so I thought the initial `cli`/`steps` vectors needed to be clean from the beginning.

I'll take another look at this approach though.

File: src/render.rs
Comment (mkmik) [inline src/render.rs:70]: we should honour the `DOCKER_CONFIG` env var and default to `format!("{}/.docker", user_home.display()` if the DOCKER_CONFIG env var is not set

File: src/apply.rs
Comment (mkmik) [reply src/apply.rs:35]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cd8ef11db1844f331f71c2e5a20d9dde

Comment (jdockerty) [reply src/apply.rs:35]: This makes sense now upon revisiting it. Error was in my brain, not in the code :laughing: 

Splitting it out between common args, with different invocation methods (`docker` when local vs `kubecfg` in controller) makes a lot more sense.

File: src/local.rs
Comment (mkmik) [inline src/local.rs:124]: I'm very surprised you need this.

Comment (jdockerty) [reply src/local.rs:124]: This was me following the compiler for a WIP commit, but it does run into another error because this is unstable.

```
error[E0658]: use of unstable library feature 'allocator_api'
   --> src/local.rs:124:35
    |
124 | ) -> Box<dyn Future<Output = ()>, Result<()>> {
    |                                   ^^^^^^^^^^
```

Comment (mkmik) [reply src/local.rs:124]: the right thing would be
```
Box<dyn Future<Output = Result<()>>> {
```

but that's not necessary because you marked the fn as `async`, and that's what happens under the hood

Comment (jeffreyssmith2nd) [reply src/local.rs:124]: I created a [PR](https://github.com/kubecfg/kubit/pull/177) against this branch that fixes this. The issue was related to recursive async calls [E0733](https://doc.rust-lang.org/error_codes/E0733.html). I just refactored the code (probably could be done better) to prevent the need for recursion.

Comment (jdockerty) [reply src/local.rs:124]: I think I understand this.

@jeffreyssmith2nd If you don't mind, can I try digest what you did in that PR and incorporate it into this one on Monday?

Although if this feature is needed urgently then it's no problem to close this one out in favour of yours so we can push it through.

Comment (mkmik) [reply src/local.rs:124]: If the PR is against this branch it can be merged here

Comment (jdockerty) [reply src/local.rs]: You're right, I missed that completely when skimming on mobile.

Massive thanks @jeffreyssmith2nd :rocket:

File: src/render.rs
Comment (jdockerty) [inline src/render.rs:7]: I initially had this in `lib.rs`, but I'm not sure whether that is the right thing to do.

I've moved it here since it makes sense for having the `kubecfg` image within the `render` portion, i.e. `kubecfg show`, but that can easily change.

Open to suggestions.

File: src/render.rs
Comment (jdockerty) [inline src/render.rs:67]: This _seems_ like the right thing to do, but maybe there's a better way.

The need for a temporary file is to leverage the pre-existing `fetch_package_config` from the `metadata` crate, rather than having our own implementation of it, I don't think that is particularly necessary here since we simply want to extract the kubecfg image/version out of it.

**Review by jdockerty — COMMENTED** [review]
I think this is a good check point for review, much appreciated for the unblock @jeffreyssmith2nd 

---

What I currently have "works", but I'm open to suggestions on whether to change bits of it too.

There is a distinct lack of testing as well, but I'm also conscious this is really just running `docker run <...>`, so I'm not sure how worthwhile that really is - perhaps its a separate thing too.

```bash
DOCKER_CONFIG="/tmp/.docker" cargo run -- local apply <path/to/overlay.yml> --dry-run script
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `target/debug/kubit local apply <path/to/overlay.yml> --dry-run script`
#!/bin/bash
set -euo pipefail

docker \
    run \
    --rm \
    -v \
    /home/influx/.kube/config:/.kube/config \
   ...remainder of script snipped
```

And running the actual `local apply` itself produces the expected output

```bash
DOCKER_CONFIG="/tmp/.docker" cargo run -- local apply <path/to/overlay.yml>
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `target/debug/kubit local apply <path/to/overlay.yml>`

deployment.apps/account serverside-applied
poddisruptionbudget.policy/account serverside-applied
service/account serverside-applied
...
```

File: src/render.rs
Comment (mkmik) [reply src/render.rs:67]: the `fetch_package_config` method does two things:

1. reads and decodes a file into an `AppInstance` structure
2. calls oci:fetch_package_config with that `AppInstance` structure and some local docker auth

Here we already have an `AppInstance` struct and we just want to leverage the fact that the `metadata::fetch_package_config` function deals with the details around handling local docker auth.

I suggest we make a little refactoring of the `metadata::fetch_package_config` function and split it into two parts, so we can skip the part where the file is read from the filesystem.

https://github.com/kubecfg/kubit/pull/179

once that's merged you can rebase this PR on top of that and make use of that function

**Review by jeffreyssmith2nd — APPROVED** [review]

**Comment by jdockerty** [comment]
I believe there's an error being introduced here with how I've altered the `emit_commandline` flow for `render.rs`, I'm looking into that now :mag: 

I'll hold off merging until I can confirm that nothing is going to start panicking when this is merged.

**Pull request merged**
